### ISSUE TITLE
Dependency Graph Output Enhancment

### ISSUE THREAD

**Issue by nathanhaigh** [hpcugent/easybuild-framework#1545]
Depenedency graph files can be generated for an easyconfig file using something like:

```bash
eb --dep-graph tool1-version-toolchain-toolchain_version.dot -r
```

The resulting `.dot` file contains a directed graph (`digraph`) defining `nodes` and `directed edges` between `nodes`.

Lines defining a `node` look like this:

```
"tool_1/version-toolchain-toolchain_version";
```

Lines defining `directed edges` look like this:

```
"tool_1/version-toolchain-toolchain_version" -> "tool_2/version-toolchain-toolchain_version";
```

This output could be enhanced by:

  * Distinguish between `build` dependencies and `normal` dependencies.

This could be done using `edge attributes`. For example, the use of `label` could be used to add text (e.g. `build dep`) to an edge. Thus a line in the `.dot` file defining a build dependency could look like:

```dot
digraph graphname {
// node definitions
"tool_1/version-toolchain-toolchain_version";
"tool_2/version-toolchain-toolchain_version"
"tool_3/version-toolchain-toolchain_version"

// edge definition for build dependency
"tool_1/version-toolchain-toolchain_version" -> "tool_2/version-toolchain-toolchain_version" [label="build dep"];
// edge definition for normal dependency
"tool_1/version-toolchain-toolchain_version" -> "tool_3/version-toolchain-toolchain_version"
}
```

Alternatives, to keep visual clutter to a minimum, it might be best to change the `style` or `colour` of the edge; or even the style of the `arrowhead`. The following example, illustrates the use of all these:

```dot
digraph graphname {
// node definitions
"tool_1/version-toolchain-toolchain_version";
"tool_2/version-toolchain-toolchain_version"
"tool_3/version-toolchain-toolchain_version"

// edge definition for build dependency
"tool_1/version-toolchain-toolchain_version" -> "tool_2/version-toolchain-toolchain_version" [label="build dep",style="dotted",color="blue",arrowhead="diamond"];
// edge definition for normal dependency
"tool_1/version-toolchain-toolchain_version" -> "tool_3/version-toolchain-toolchain_version"
}
```

**Comment by nathanhaigh** [comment]
Useful resources:

  * https://en.wikipedia.org/wiki/DOT_%28graph_description_language%29#Attributes
  * http://www.graphviz.org/doc/info/attrs.html
  * http://www.graphviz.org/pdf/dotguide.pdf

**Comment by boegel** [comment]
Thanks for the detailed overview @nathanhaigh!

To whomever wants to look into this: you'll need to play around with the `dep_graph` function in the `easybuild.framework.easyconfig.tools` module.

The `.dot` file is generated via the `dot.write()` function provided by the python-graph-dot package.

Whether @nathanhaigh's suggestions can be implemented easily depends on whether python-graph-dot supports these type of tinkering with edges and labels...

**Comment by nathanhaigh** [comment]
According to http://pydoc.net/Python/python-graph-core/1.8.0/pygraph.classes.digraph/ edge attributes can be added like so:

```python
add_edge(edge, wt = 1, label="", attrs = []):
```

Where `attrs` is:
> @type  attrs: list
   @param attrs: List of node attributes specified as (attribute, value) tuples.

So @boegel, I guess I just need to know what's available via `specs`.

**Comment by boegel** [comment]
@nathanhaigh: it looks like you should be able to check with `dep.get('build_only', False)` where `add_edge` is being used to figure out whether `dep` is a build dependency

**Comment by boegel** [comment]
see https://github.com/hpcugent/easybuild-framework/blob/master/easybuild/framework/easyconfig/easyconfig.py#L619

**Comment by boegel** [comment]
fixed in #1548, thanks a lot @nathanhaigh for tackling this!
